nltools/data/brain_data.py

```diff
@@ -1205,26 +1205,17 @@ def apply_mask(self, mask):
             masked: (Brain_Data) masked Brain_Data object
 
         """
-
-        if isinstance(mask, Brain_Data):
-            mask = mask.to_nifti()  # convert to nibabel
-
-        if not isinstance(mask, nib.Nifti1Image):
-            if isinstance(mask, six.string_types):
-                if os.path.isfile(mask):
-                    mask = nib.load(mask)
-                if not ((self.mask.get_affine() == mask.get_affine()).all()) & (self.mask.shape[0:3] == mask.shape[0:3]):
-                    mask = resample_img(mask, target_affine=self.mask.get_affine(), target_shape=self.mask.shape)
-            else:
-                raise ValueError("Mask is not a nibabel instance, Brain_Data "
-                                 "instance, or a valid file name.")
-
+        mask = check_brain_data(mask)
         masked = deepcopy(self)
-        nifti_masker = NiftiMasker(mask_img=mask)
-        masked.data = nifti_masker.fit_transform(self.to_nifti())
-        masked.nifti_masker = nifti_masker
-        if (len(masked.shape()) > 1) & (masked.shape()[0] == 1):
-            masked.data = masked.data.flatten()
+
+        if masked.shape()[0] == mask.shape()[0]:
+            masked.data[~mask.data] = 0
+        else:
+            nifti_masker = NiftiMasker(mask_img=mask)
```

Comment:
Ok so this is converting the original `Brain_Data` instance into whatever space the `mask` is in by creating a `NiftiMasker` out of the `mask`. Should we not be doing the opposite? Using the original `Brain_Data`'s mask to transform the mask we want to apply? Otherwise it's confusing that you could be changing spaces without realizing it.

A second issue is that this will give a different return object compared to how you're dealing with 1d brains on like 1211 above. Above you're _masking out_ non-mask values by setting them to 0 but retaining the original 1d shape. In this solution you could get back fewer voxels (e.g. an ROI with 100 voxels) and thereby get a different return shape. 

The function should be consistent with whether we want to always return back something of the same shape as the original, but with 0'd out voxels, or a subset of the original data shape of the same size as the input mask. I raised this point in issue https://github.com/cosanlab/nltools/issues/304. There are pros and cons for both and our decision could break a lot of existing lab scripts

Reply:
> Ok so this is converting the original `Brain_Data` instance into whatever space the `mask` is in by creating a `NiftiMasker` out of the `mask`. Should we not be doing the opposite? Using the original `Brain_Data`'s mask to transform the mask we want to apply? Otherwise it's confusing that you could be changing spaces without realizing it.
> 

This is the intended use case for this function to convert the data into the mask space. I'm now realizing that the logical indexing approach can return different results, which is problematic. Trying to think about how to solve this.

> A second issue is that this will give a different return object compared to how you're dealing with 1d brains on like 1211 above. Above you're _masking out_ non-mask values by setting them to 0 but retaining the original 1d shape. In this solution you could get back fewer voxels (e.g. an ROI with 100 voxels) and thereby get a different return shape.


> The function should be consistent with whether we want to always return back something of the same shape as the original, but with 0'd out voxels, or a subset of the original data shape of the same size as the input mask. I raised this point in issue #304. There are pros and cons for both and our decision could break a lot of existing lab scripts

Agreed, that this is not desirable for this function. Here I think we want to return the subset of the voxels with the same space as the input mask. The main issue is that we need to use the nifti_masker transform of the mask, otherwise this can create problems transforming back into 3D space.